src/nvim/runtime.c

```diff
@@ -2177,28 +2144,32 @@ int do_source(char *fname, int check_other, int is_vimrc, int *ret_sid)
 
   const sctx_T save_current_sctx = current_sctx;
 
-  current_sctx.sc_lnum = 0;
+  if (!cookie.source_str_from_lua) {
+    current_sctx.sc_lnum = 0;
+  }
 
   // Always use a new sequence number.
   current_sctx.sc_seq = ++last_current_SID_seq;
 
   if (sid > 0) {
     // loading the same script again
     si = SCRIPT_ITEM(sid);
-  } else {
+  } else if (cmd == NULL) {
     // It's new, generate a new SID.
     si = new_script_item(fname_exp, &sid);
     fname_exp = xstrdup(si->sn_name);  // used for autocmd
     if (ret_sid != NULL) {
       *ret_sid = sid;
     }
   }
-  current_sctx.sc_sid = sid;
+  if (!cookie.source_str_from_lua) {
+    current_sctx.sc_sid = sid;
+  }
```

Comment:
Probably better to make `nvim_exec2` behavior the same in Vimscript and Lua (and not allocated a script ID unless using `:s`), while always allocating a new script ID when sourcing buffer lines.